org/mozilla/jss/ssl/javax/JSSServerSocketFactory.java

```diff
@@ -0,0 +1,90 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.io.*;
+import java.net.*;
+import java.security.*;
+
+import javax.net.ssl.*;
+
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.ssl.SSLCipher;
+
+public class JSSServerSocketFactory extends SSLServerSocketFactory {
+    private SSLContext ctx;
+    private JSSKeyManager key_manager;
+    private X509TrustManager[] trust_managers;
+
+    public JSSServerSocketFactory(String protocol, JSSKeyManager km, X509TrustManager[] tms) {
```

Comment:
Happy to explain what I know :)

Let's start with the keys and work up:

1. In Java, there's a [KeyStore](https://docs.oracle.com/javase/8/docs/api/java/security/KeyStore.html). This blindly stores (public/private/symmetric) keys and certificates. Typically, under traditional Java there's several of these, and a key store matching each type:
   1. A NSS DB under `SunPKCS11-NSS{,-FIPS}`,
   2. A traditional "java key store (JKS)",
   3. A .p12 file,
   4. "flat" PEM files
 - A `KeyStore` is constructed _from_ the provider itself. E.g., the `SunPKCS11-NSS` provider only exposes a NSS DB-type key store. You get an instance (e.g., with name `PKCS12`) and then can `.load()` and `.store()` to/from a Stream/properties class.
2. The _Java security provider_ provides a way of mapping `KeyStore`->`KeyManager` via a `KeyManagerFactory`. Typically this needs to be initialized with a specific `KeyStore` instance we wish to use. You can then call `getKeyManagers()` and all possible `KeyManager`s from that `KeyStore` are created.
3. There's really only one "type" of [`KeyManager`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/KeyManager.html) right now: a [`X509KeyManager`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/X509KeyManager.html) and its only implementing abstract class, the [`X509ExtendedKeyManager`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/X509ExtendedKeyManager.html). However, there could be e.g., a `SymmetricKeyManager` that selects which key is best used for, e.g., a SSH connection. I don't know. 

OK so back to JSS:
 - We never had a KeyStore implementation. This has historically been filled by CryptoManager.
 - Our KeyManagerFactory didn't need initialization since CryptoManager is a singleton.
 - KeyManager and TrustManager were originally written by you, for TomcatJSS and later migrated into JSS. There's still only a single source of keys, which is well, CryptoManager.
 - Tomcat will wrap [a KeyManager](https://github.com/apache/tomcat/blob/master/java/org/apache/tomcat/util/net/SSLUtilBase.java#L376) instance and [only return a single key as specified in the config](https://github.com/apache/tomcat/blob/master/java/org/apache/tomcat/util/net/jsse/JSSEKeyManager.java#L29-L35).


So for the foreseeable future, we really only have/need a single JSSKeyManager implementation. In the future, if we wish to support more KeyManagers, we could:
 - Potentially supporting NSS interfaces to load additional [Cert DBs](https://github.com/nss-dev/nss/blob/master/lib/certdb/cert.h#L362) and expose the result as a `KeyStore`
 - Provide P12 and PEM parsing via JSS (including loading their data into temporary cert/key storage) and expose the result as a `KeyStore`,
 - _Potentially_ make CryptoManager not a Singleton (though unlikely)
 - Make KeyManagerFactory support and respect a `KeyStore` parameter.
 - Introduce real KeyManager classes over the top of those KeyStores.
 - We don't really support non-JSS KeyManagers within our JSSSocket / JSSEngine implementations because we don't have a 

But for now, most of that isn't necessary. We've been fine using our own PKCS12/PEM parsing logic in Dogtag PKI (so KeyStore doesn't get us much), we'd have to figure out how to import keys in FIPS mode (which sometimes isn't allowed anyways). And probably a lot of other stuff I'm not aware of. :)

That make sense, @edewata?

Reply:
Thanks for the explanation!

Just FYI, we do have a `JSSKeyStoreSpi`. And yes, `CryptoManager` will probably remain a singleton for foreseeable future, but I think JSS would be more useful in general if we can remove that limitation. IIUC NSS itself doesn't have this limitation, so ideally JSS shouldn't either.

So IIUC right now we only have/use a single key manager. I just thought that since Java API in general defines methods with multiple key managers, we probably should do the same (even though internally we only use one). But this will not be an issue any time soon, so no worries. Thanks!

Reply:
@edewata  -- Ah so I am wrong. :-) JSS _does_ expose a JSSKeyStoreSpi implementation! I love when code appears without me having to write it ;-) And -- we've both also modified this code and forgotten about having done so! 

So then, when you wrote the `JSSKeyManager` / `JSSTrustManager` implementations originally, they should've taken a `KeyStore` instance as a parameter and perhaps validated that it was an instance from the `Mozilla-JSS` provider... When I added the `JSSKeyManagerFactory` and `JSSTrustManagerFactory`, they too should've respected the `KeyStore` parameter... But we haven't done any of that.

I think it'd be simple enough to add if we want to. Thoughts?

Reply:
Sure, if you have time for it why not? :) Thanks!